Add basic unit tests

hershen opened the pull request:
Meant to be reviewed after #6.

hershen committed:
Rename eval function to evaluate

hershen committed:
Remove hersh/hershe.test.cpp

Remove that file as it was only used to make sure Catch2 was working,
and was not testing anything.

hershen committed:
Change how printer::print works

printer::print returns the line to be printed, instead of sending it to
std::cout. The main REPL loop is now responsible to send the std::string
returned by printer::print to std::cout.

hershen committed:
Change the reader::read function into a Reader class

The Reader constructor can be used to set the std::istream from which the
input is read and the std::ostream to which the prompt is sent. This
makes it easier to unit test, as both the istream and ostream (and their
content) can be controlled by the testing function.
The class can also be used to set the prompt.

Also remove the " " that was added by the reader::read after the prompt. This " " can now be part of the prompt set in the Reader class.

hershen committed:
Add basic tests for reader::Reader, printer::print and evaluate::evaluate

hershen committed:
Re-arange include order in test files

Re-arange according to
https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes.

oorijsonar commented on hersh/evaluate.test.cpp:
```diff
@@ -0,0 +1,9 @@
+
+TEST_CASE("Test evaluate", "[evaluate]") {
+    
+    REQUIRE(hersh::evaluate("") == "");
+    REQUIRE(hersh::evaluate("something ") == "something ");
```
These should probably separate sections. Otherwise I'm not sure why we need both.

hershen replied:
I like to always test the empty case.

oorijsonar replied:
That means you think it's a special case therefore it should have its own section.

hershen replied:
Partitioning into SECTIONS causes everything between TEST_CASE and the first SECTION to be run before each SECTION. Like a set-up. I don't think there's a need for set-ups here (yet?).

And why should special cases be in a different section even if there's no set-up issues?

oorijsonar replied:
I'm used to organizing tests in individual test cases and it seems like you're using sections as if they separate test cases. Adding addition assertions (`REQUIRES` in this case) should be done for each test case individually. The test case's name indicates what you're testing.
Therefore having a test case named "check empty case" and one named "check happy path" would be more useful in this case. That way when a test case fails you understand what it's supposed to test immediately by reading its name. I don't know if they should be separate `SECTION`s or separate `TEST_CASE`s but they should be separate something.

hershen replied:
I think that makes sense. I was following the example in the [Catch2 tutorial](https://github.com/catchorg/Catch2/blob/devel/docs/tutorial.md#test-cases-and-sections) is not a good example?
Are you saying it's not a good example?

Edit: I think I understand - partition with either TEST_CASE or with SECTION.

oorijsonar replied:
👍

hershen replied:
Done. There's no nested TEST_CASE, so you were right, it needs to be partitioned with SECTIONs.

oorijsonar commented on hersh/hersh.cpp:
```diff
@@ -6,23 +6,26 @@
 using namespace hersh;
 
-const std::string defaultPrompt = "hersh";
+const std::string defaultPrompt = "hersh$ ";
```
Shouldn't the default prompt live in the `reader`?

hershen replied:
Yes, that's better.

oorijsonar commented on hersh/hersh.cpp:
```diff
@@ -6,23 +6,26 @@
 
   while(true) {
-    const auto readLine = read(defaultPrompt);
+    const auto readLine = reader.read();
```
like

oorijsonar commented on hersh/hersh.cpp:
```diff
@@ -6,23 +6,26 @@
 }
 
 int main(int argc, char *argv[]) {
+  Reader reader;
```
Should probably be initialized with a prompt. A `reader` without a prompt is like a chimney sweep without a chimney sweeper.

hershen replied:
I set Reader's default prompt to "hersh$ ". So no need to initialize. Or is that what you meant?

oorijsonar replied:
Partly. It's better now. But, it doesn't really make sense to set the prompt after initializing a `Reader` so I would get rid of the setter and add a constructor that allows initializing it.

hershen replied:
I haven't read the whole sh specifications, but if it's similar to the default bash prompt behavior, the prompt changes each time you change directory. So I think we'll have to change the prompt a lot.

oorijsonar replied:
👍

oorijsonar commented on hersh/printer.test.cpp:
```diff
@@ -0,0 +1,9 @@
+
+TEST_CASE("Test printer", "[printer]") {
+    
+    REQUIRE(hersh::print("") == "");
+    REQUIRE(hersh::print("something ") == "something ");
```
Same as above.

hershen replied:
Done.

oorijsonar commented on hersh/hersh.cpp:
```diff
@@ -6,23 +6,26 @@
 
-    const auto evaluatedLine = eval(readLine);
-    print(evaluatedLine);
+    const auto evaluatedLine = evaluate(readLine);
```
like

oorijsonar commented on hersh/reader.hpp:
```diff
@@ -3,18 +3,45 @@
-  std::string line;
+class Reader {
+public:
+  explicit Reader() : 
```
Can you explain why use `explicit` here? I've never seen it before

hershen replied:
It prevents the constructor from doing implicit conversions. 
See explanation in [cppreference](https://en.cppreference.com/w/cpp/language/explicit) and the examples at the bottom which are illuminating, and also [stackoverflow](https://stackoverflow.com/questions/121162/what-does-the-explicit-keyword-mean). They have good explanations. I've never used it before, but it's intended to prevent unintended and hard to find bugs. We can try it see if it's useful.

You might say that it's not needed for empty constructors, and that's true. But I have a feeling it's a good habit to make every constructor explicit.

oorijsonar commented on hersh/reader.test.cpp:
```diff
@@ -0,0 +1,29 @@
+
+    hersh::Reader reader(istringstream, ostringstream);
+
+    REQUIRE(reader.getPrompt() == "");
```
I'm not sure this is a behavior worth testing... Is it important to us that the default prompt is an empty string?

hershen replied:
I think it's important to test that the class does what you think it does. If you think the default prompt is empty, you should check that. That's how I see it.

oorijsonar replied:
That means that every time we change the default prompt we need to change the test. I think this test causes more problems than it solves.

hershen replied:
Yes, you'd have to change the test, because you change the class. I don't see a problem with that. 

In any case, I don't think we'll change the default prompt a lot. Maybe one more time to make it more consistent with bash ( just '$' instead of 'hersh$').

oorijsonar replied:
When is this test useful? When you make a typo in the default prompt? It causes trouble and will never catch any bugs.

hershen replied:
Done.

hershen committed:
Merge branch 'main' into addUnitTests

hershen committed:
Additions to previous merge commit, after changing project structure

1. Move unit test files into src/ directory.
2. Base include directory for compilation is now src/.

hershen committed:
Maker Reader's default prompt "hersh$ "

hershen committed:
Partition tests with SECTIONs

Partition tests with SECTIONs in printer.test.cpp and evaluate.test.cpp.

hershen committed:
Remove test of Reader default prompt

Remove the test as it only tests the content of Reader's private member
variable.

hershen committed:
Change test

Remove the hard coded prompt from the test and replace with
Reader::getPrompt(), as the default prompt may change.

hershen commented:
All comments addressed. Waiting for re-approval.
